src/webgpu/shader/validation/expression/call/builtin/quantizeToF16.spec.ts

```diff
@@ -0,0 +1,104 @@
+const builtin = 'quantizeToF16';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { Type, kConcreteF32ScalarsAndVectors } from '../../../../../util/conversion.js';
+import { quantizeToF16 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord([
+  Type.abstractFloat,
+  Type.vec(2, Type.abstractFloat),
+  Type.vec(3, Type.abstractFloat),
+  Type.vec(4, Type.abstractFloat),
+  ...kConcreteF32ScalarsAndVectors,
+]);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() never errors
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('value', u => fullRangeForType(kValidArgumentTypes[u.type]))
+  )
+  .fn(t => {
+    let expectedResult = true;
+
+    // Should be invalid if the quantized value exceeds the maximum representable
+    // 16-bit float value.
+    const f16Value = quantizeToF16(Number(t.params.value));
+    if (f16Value === Infinity || f16Value === -Infinity) {
+      expectedResult = false;
+    }
+
+    const type = kValidArgumentTypes[t.params.type];
+
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [type.create(t.params.value)],
+      t.params.stage
+    );
+  });
+
+const kArgCases = {
+  good: '(vec3f())',
+  bad_no_parens: '',
+  // Bad number of args
+  bad_0args: '()',
+  bad_2arg: '(1.0, 2.0)',
+  // Bad value for arg 0
+  bad_0bool: '(false)',
+  bad_0array: '(array(1.1,2.2))',
+  bad_0struct: '(modf(2.2))',
+  bad_0f16: '(1h)',
```

Comment:
This is interesting. It can fail for two reasons:
shader-f16 not supported, so the literal doesn't parse.
Or the overload is not available.

So this is correct, for a subtle reason.

Reply:
Good point! Updated this to be less "accidentally correct". :)